src/OrchardCore.Modules/OrchardCore.Localization/TagHelpers/CulturePickerTagHelper.cs

```diff
@@ -0,0 +1,66 @@
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Localization;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using OrchardCore.ResourceManagement;
+using OrchardCore.Settings;
+
+namespace OrchardCore.Localization.TagHelpers
+{
+    [HtmlTargetElement("culture-picker", TagStructure = TagStructure.WithoutEndTag)]
+    public class CulturePickerTagHelper : TagHelper
+    {
+        private readonly ISiteService _siteService;
+        private readonly IResourceManager _resourceManager;
+
+        public CulturePickerTagHelper(ISiteService siteService, IResourceManager resourceManager)
+        {
+            _siteService = siteService;
+            _resourceManager = resourceManager;
+        }
+
+        [ViewContext, HtmlAttributeNotBound]
+        public ViewContext ViewContext { get; set; }
+
+        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+        {
+            output.TagName = null;
+
+            var selectedCulture = ViewContext.HttpContext.Features.Get<IRequestCultureFeature>().RequestCulture.Culture.Name;
+            var siteSettings = await _siteService.GetSiteSettingsAsync();
+            var selectTag = new TagBuilder("select");
+            selectTag.AddCssClass("form-control");
+            selectTag.Attributes.Add("onchange", "setLocalizationCookie(this.value))");
+
+            foreach (var culture in siteSettings.SupportedCultures)
```

Comment:
Default culture could be different than siteSettings.SupportedCultures so it would be missing. You could easily do a foreach on the current Localization middleware supported cultures instead. No need to look at site settings here.

Reply:
```C#
var supportedUICultures = context.HttpContext.RequestServices.GetService<IOptions<RequestLocalizationOptions>>().Value.SupportedUICultures;
```

Reply:
If the Localization Middleware is disabled then siteSettings.SupportedCultures will return cultures that are not supported by the Middleware. If the Localization Middleware is disabled the only culture supported will be the current system one. So that's why on a culture picker for frontend website we should use this code line I sent you previously.

Reply:
OK got your point .. thanks